### ISSUE TITLE
Text selection stops working

### ISSUE THREAD

**Issue by marilmanen** [eteran/nedit-ng#140]
This is a strange issue and I have not been able to find a way to reproduce it, but it seems to happen  when I reconnect to my remote desktop.

Sometimes I loose possibility to select any text from the nedit-ng edit area.  Selection can't be done with mouse or with Shift+arrow keys. The issue is related to existing nedit-ng windows only i.e. selection in terminal and other applications is still working correctly. Any new nedit-ng window that I start works also correctly. I have not found any way to enable selection on a broken nedit-ng window.

**Comment by eteran** [comment]
Very strange indeed. Since you can't reproduce it, does that mean that it's only happened once? Or has this happened a few times?

Also, did other forms of input work like typing and moving the cursor around? Was it just selection that stopped working, or all mouse input for that window?

I'll see if I can get to the bottom of it if we can narrow it down slightly.

**Comment by marilmanen** [comment]
It has happened about 5 to 10 times, so I thought it's worth reporting. Everything else works, but now that I think it, I'm not sure if it's just a highlight issue. Selecting text with mouse moves the cursor to the selection end point and maybe the selected text is just not marked as selected. I'll need to double check it, but my feeling is that it is a selection issue and not just a highlight issue.

**Comment by eteran** [comment]
OK, that's a very strange issue indeed. I'll keep an eye out for it

**Comment by anjohnson** [comment]
I have seen this a couple of times on RHEL 7.7, after I've had a single nc-ng session running for many days (weeks maybe), and possibly a weekend's inactivity too. It required shutting down that session and starting a new server, I couldn't get text selection working again in the broken session.

**Comment by rnhaddad** [comment]
Happens very frequently in fact. To resolve, i have to re-open some files.

**Comment by eteran** [comment]
@anjohnson @rnhaddad Are you both also using a remote desktop of some kind? Also, what distro are you guys using, it makes me wonder if it's a Qt bug you're encountering... I dunno.

I've never had this happen to me, so I'm not quite sure where to start when looking into it yet, but I'll see what I can come up with.

**Comment by anjohnson** [comment]
I am not using a remote desktop; my setup is a pair of monitors connected to a single NVIDIA Quadro 2000 card with the X11 server on RHEL-7.7, Linux 3.10.0-862.6.3.el7.x86_64 with `DISPLAY=:0`. The libQt* RPMs that nedit-ng is linking against are:
```
qt5-qtbase-5.9.7-2.el7.x86_64
qt5-qtx11extras-5.9.7-1.el7.x86_64
qt5-qtbase-gui-5.9.7-2.el7.x86_64
```

I don't get these symptoms very often and haven't seen them in several months now so I can't be sure if they are still happening. It sounds like @rnhaddad might be more useful than me in tracking this down.

**Comment by eteran** [comment]
@rnhaddad @anjohnson have either of you experienced this recently? I've never been able to reproduce this locally :-(.

**Comment by anjohnson** [comment]
I haven't been using nedit-ng continuously on my RHEL-7 box much since $WORK went remote back in March. I have never seen this on the MacOS laptop I have at home, but that device can't seem to stay alive for weeks on end like that workstation does, so I can't confirm whether it still happens or not – sorry!

**Comment by rnhaddad** [comment]
@eteran still happening. Any new changes you want me to try out (I am still using a relatively old SHA).

**Comment by eteran** [comment]
I certainly like to think that it has been improved a bunch :-).

Mostly it's been a lot of internal stuff and bug fixes though. I made a big fix to the macro system recently. Depending on how old your version is, I may or may not have inadvertently fixed this issue by refactoring other code, so if you're interested in doing so, I'd love to know if you still experience this issue.

Thanks!

**Comment by marilmanen** [comment]
I have been using version from April and with it I have seen issues with text selection. I'm currently updating to the latest version and let you know in case I can still see the issue with text selection.

**Comment by dherringtn** [comment]
Hi @eteran,

I just downloaded and compiled nedit-ng and started using it last week since I kept having problems with nedit-5.6 copy-paste functionality breaking. Thanks for such a great update to nedit! It has a great look and feel!

Unfortunately I'm seeing the issue that @marilmanen posted, where the text selection stops working in an nedit-ng window. The selection flashes off and on rapidly as I try to drag the mouse over some lines of text.

I'm connected to a RedHat 8 machine through VNC if that matters. Another interesting thing about selecting text is that if I select several lines of a script that I'm viewing in nedit-ng, and then paste it onto the command line of a terminal, when I go back to the nedit-ng window and scroll down with my mouse, the selection incorrectly follows my mouse direction and highlights more text without me even pressing the button!

I even crashed my VNC viewer session at one point when I was dragging to select and/or scrolling, and had to re-connect to my VNC session.

I'm wondering if there's anything I can try to resolve this, or maybe what info I can look for to try to debug what's going on.
Thanks!

**Comment by eteran** [comment]
@dherringtn Hmm, that is unfortunate! So I have a couple of thoughts/questions:

1. Which version have you tried? Are you using the latest release, master, one of the others? If you haven't tried master, then it's worth a go, I have a (bad) habit of going long stretches between releases despite updating a lot of code.

2. I fear that this may be a bad interaction with VNC, especially the bit about switching windows and it extending the selection without you clicking. That sounds an aweul lot like Qt missed a button press/release event. (Basically, if Qt/Nedit sees a mouse down and then never sees the mouse up, well, it'll assume you're still holding down the mouse button forever ... or more realistically, until the NEXT time you cause a mouse up event by clicking again).

3. Is it possible for you to test usage without VNC in the picture? The fact that VNC has even crashed is to me concerning that it may be outside of my control. Because nedit-ng, of course doesn't do anything that should effect VNC's functionality. It doesn't even know VNC exists.

**Comment by anjohnson** [comment]
@dherringtn Have you tried disconnecting and reconnecting your VNC viewer session when it gets into that state? I have very occasionally had issues with an NX (NoMachine) remote session from a Mac where nedit-ng on Linux got confused about a key state. This seemed to be caused by a key-up event from the Alt or Shift key that somehow got lost through the session transport (or maybe a modal dialog that popped up somehow stole that keyboard event). Reconnecting to the NX session cleared the key state and restores proper operation of the selection state in nedit-ng.

**Comment by eteran** [comment]
@dherringtn If you don't care that if you rebuild it next week you might get updated code, then I'd just stick with master, it's where the "best" version of the code lives (for some definition of best ;-) ).

Regarding the multi-desktop/new-window stuff. That code should be in the mainline (and has been for at least a couple of releases). If you're curious and are a C++ programmer, it's in `NeditServer.cpp`. The function `FindDocumentOnDesktop` attempts to find a candidate window to place the new editor in on the desktop is the "current desktop". Which I define as the one which currently has the cursor.

Unfortunately, I worry that maybe VNC does something that Qt doesn't understand regarding multi-desktops. I don't mean to just blame VNC, as It also could be that Qt has an implementation that is weak under VNC as well! But really the bottom line is that nedit-ng, unlike classic nedit, does not do any raw X11 stuff, at all. So I'm at the mercy of what VNC exposes and what Qt thinks it sees.

One thing that's worth exploring is that master, unlike the last release, can build with much newer versions of Qt. It's worth seeing if you can get Qt 6.7+ installed, maybe even 6.9 and seeing if that works any better for you. Qt 5 (which is what I assume you're using), is pretty old at this point

let me know if any experiments help (or don't).

And BTW, it occurs to me, if you're using VNC linux <-> linux, have you considered just using a remote X display instead? it's likely to work much smoother.

**Comment by marilmanen** [comment]
There seems to be an issue with text selection starting from version a2f869f9 or at least I can't reproduce it with versions c51fcc00 or
bbe581b1

This is how I can see an issue. 
1. open new nedit-ng and type some text or open a file that contains text, but make sure that there's so much text that a vertical scroll bar is visible.
2. select one or more lines of the text
3. activate some other window with mouse left button on your workspace
4. click with left mouse button on the title bar of your nedit-ng window
5. scroll with mouse wheel and the selection area is changing and that's not what should happen

**Comment by eteran** [comment]
@marilmanen I'll have to look into that, you always do find the fringe cases!

Since https://github.com/eteran/nedit-ng/commit/a2f869f9503afb0130013a1e6b10f12b0741ab49, (IIRC) I had to manually implement some click tracking to resolve the bigger issue you were experiencing. I have a few theories about what could be happening, I'll let you know what I find.

**Comment by eteran** [comment]
@dherringtn OK, I have reproduced at least part of your bug and it's definitely a bug. Steps to reproduce "phantom scrolling":

1. Open a particularly large doc (to give it time to happen)
2. Click in side the window
3. Still holding the mouse button, move below the window
4. Release the button

observe it still scrolls basically forever until your next click.

So the issue seems to be that it doesn't see the Mouse Release event in this case. The strange thing is this:

There's a timer in the code to repeat the scrolling as needed. But if I simply **disable** that timer, things work basically as I'd expect. I need to experiement, but this may be basically a bug that's always been in my port due to Qt handling this already and it not needing the timer at all.

I'm going to create a new issue referencing this.

@marilmanen I'll do the same with your new observations.

**Comment by eteran** [comment]
Actually I think @marilmanen 's observations are the SAME bug (it has not "always been there"). So it'll go in one ticket.

**Comment by eteran** [comment]
@marilmanen @dherringtn I just pushed a fix for the "ghost scrolling" thing that was happening. Maybe this will also help with some of the VNC weirdness?

**Comment by dherringtn** [comment]
@eteran Thanks! I'll try the latest and see how it goes with the fix. Also, it looks like I may be stuck with Qt 5.15.3 on this machine unless I can get IT to install a newer version.

**Comment by eteran** [comment]
Yeah in principle, Qt5 is still great. I only recently made QT6 usable at all 🤣
